Admin Grid Mass action Select / Unselect All issue #9610

minesh0111 opened the pull request:
Admin Grid Mass Action `Select All ` and `Unselect All` work only on `primary_key` of collection object no matter if we set other field as mass action field using grid method `setMassactionIdField` other than `primary_key`

### Preconditions
1. Magento version 2.1.3


### Steps to reproduce
<!--- Provide a set of unambiguous steps to reproduce this bug include code, if relevant  -->
1. Found this issue while creating custom module where i need mass action id field other than `primary_key` of collection. 

### Expected result
<!--- Tell us what should happen -->
1. It should `Select All` and `Unselect All`  on click of option
2. It should post `MassactionIdField`  on submit

### Actual result

1. Checkbox `checked` and `unchecked` not working on click of option
2. Always post  `primary_key` value of collection on submit

minesh0111 committed:
Admin Grid Mass action Select / Unselect All issue #9610

minesh0111 committed:
Merge pull request #1 from minesh0111/minesh0111-patch-1

Admin Grid Mass action Select / Unselect All issue #9610

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=9611) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=9611) before we can accept your contribution.<br/>

fooman commented:
hi @minesh0111 thanks for your contribution. Quick question with your new implementation wouldn't this result in an additional load of the ids in the case that the massaction is the primary key?

fooman commented:
There are a couple of unit test failures in relation to the changes you made as outlined [here](https://travis-ci.org/magento/magento2/jobs/231453462) - would you mind taking a look to see if you can get them updated to the new code?

madhav-primetraffic commented:
can someone guide me on The Travis CI build failed report and how to solve it ?

minesh0111 commented:
@fooman i am new to unit test can you guide me to fix that ?

convenient commented:
Have a look at the failing test class and method 

`Magento\Backend\Test\Unit\Block\Widget\Grid\Massaction\ExtendedTest::testGetGridIdsJsonWithUseSelectAll`

http://devdocs.magento.com/guides/v2.0/config-guide/cli/config-cli-subcommands-test.html

Tests can be hard to get your head around initially, but once you understand how phpunit is configured it really opens up doors and improves robustness of code.

You can see the directories specified for unit testing here: https://github.com/magento/magento2/blob/develop/dev/tests/unit/phpunit.xml.dist#L14

minesh0111 commented:
@convenient thanks for your feedback. yes i need  to understand unit test first then will go for magento unit test. 

do we need to create unit test our self or PHP unit test will do this job ?

convenient commented:
There is already a test in place covering the area you changed I believe. Your changes tripped it up as they're a change in functionality which is now causing them to fail. You'll need to

- run these test locally to see them fail
- update the tests to make them pass
- push

minesh0111 commented:
@convenient do i need to make another pull request for same ?

convenient commented:
You should push up into the same branch. All code changes to make your feature work, as well as going green in the unit tests, should be in this Pull Request. 👍

minesh0111 committed:
Travis CI Fixed For issue #9610

minesh0111 committed:
Travis CI Fixed For issue #9610

minesh0111 commented:
@convenient All checks have passed :+1:

fooman commented:
good work @minesh0111 and thanks @convenient for helping out.

@minesh0111 do you have a minimal example module that could be used to verify your changes?

minesh0111 commented:
@fooman if it is required i need to create it for sample

fooman commented:
@minesh0111 looks good - nothing further needed for now. We'll be running a few more tests internally on this and get back to you.

One question: do you see any reason why this change should not also be applied to
`Magento\Backend\Block\Widget\Grid\Massaction\AbstractMassaction` which has the same getGridIdsJson() implementation?

minesh0111 commented:
@fooman yes you are right. i will update for `Magento\Backend\Block\Widget\Grid\Massaction\AbstractMassaction` as well.

minesh0111 committed:
Admin Grid Mass action Select / Unselect All issue #9610

minesh0111 committed:
Travis CI Fixed For issue #9610

minesh0111 commented:
@fooman  Done for `Magento\Backend\Block\Widget\Grid\Massaction\AbstractMassaction`

minesh0111 committed:
Codacy Whitespace issue fixed

minesh0111 committed:
Codacy Whitespace issue fixed

minesh0111 committed:
Codacy Whitespace issue fixed

minesh0111 committed:
Codacy Whitespace issue fixed

minesh0111 committed:
codacy fixed

minesh0111 commented:
@fooman please look into i am tired now by fixing codacy

minesh0111 committed:
codacy fixed

fooman commented:
well done @minesh0111 and thanks for persevering. Will let you know if there is anything else that comes up when merging but looks great to me so far.

fooman commented:
hi @minesh0111 it unfortunately looks like the change has two unintended consequences. 

First one is on the Marketing > Reviews screen which changed from

<img width="1705" alt="screen shot 2017-05-24 at 12 42 50" src="https://cloud.githubusercontent.com/assets/455508/26382181/9d99b5c0-407e-11e7-9508-9143bc8157b1.png">

to

<img width="1708" alt="screen shot 2017-05-24 at 12 43 10" src="https://cloud.githubusercontent.com/assets/455508/26382183/9ff9c56c-407e-11e7-86ab-2ccf6db37517.png">

And the other one is during the creation of configurable products

![2017-05-23 23-37-31 14164 1d3d8ad2c7](https://cloud.githubusercontent.com/assets/455508/26382245/180f9b80-407f-11e7-9e14-ff5900c26e15.png)

The failures seem to have in common that they already use `setMassactionIdField()` (for example app/code/Magento/Review/Block/Adminhtml/Grid.php) and it currently works because that was ignored. 

Let me get someone to chime in on how this is supposed to work and what the best approach we can take in the current circumstances.

minesh0111 commented:
@fooman for review block it is showing exception in developer mode

 **Exception #0 (Zend_Db_Select_Exception): You cannot define a correlation name 'store' more than once**

fooman commented:
thanks @minesh0111  - to be able to get your PR merged we'll unfortunately need to fix all the incorrect implementations that are brought to light with your change.

I believe removing just this line 
https://github.com/magento/magento2/blob/develop/app/code/Magento/Review/Block/Adminhtml/Grid.php#L313 might do the trick for the reviews.

However we'll unfortunately need to take a look at every single use where a grid uses `setMassactionIdField()`. For now let's focus on the review and configurable product grid first and then see what other grids might require attention.

minesh0111 committed:
Removed clear() method call on collection

Removed clear() method call on collection to prevent multiple call of method _beforeLoad() on collection.

minesh0111 committed:
Removed clear() method call on collection

Removed clear() method call on collection to prevent multiple call of method _beforeLoad() on collection.

minesh0111 committed:
Travis CI Fixed For issue #9610

minesh0111 committed:
Travis CI Fixed For issue #9610

minesh0111 commented:
@fooman updated my code. waiting for checks...

minesh0111 committed:
codacy fixed

minesh0111 commented:
@fooman Please look into why it Travis CI is getting failed. It has nothing to do with my changes.

fooman commented:
thanks @minesh0111 - the travis issue seems to have been some date issue. A rerun solved it. I'll progress your PR along and keep you posted.

fooman commented:
@minesh0111 unfortunately bad news - there is another failure coming from this functional test https://github.com/magento/magento2/blob/develop/dev/tests/functional/tests/app/Magento/PageCache/Test/TestCase/CacheStatusOnScheduledIndexingTest.php with this patch in place - I have requested some more info and will post back here when I know more.

minesh0111 commented:
@fooman this commit going crazy now. can you point me on this. how can i test this functional test ?

fooman commented:
@minesh0111 I understand the frustration. You unfortunately picked a component (grids) which is used all over Magento so we need to make sure that the introduced changes don't break anything.

@ishakhsuvarov do you mind providing some more guidance on what actually failed in the mentioned test?

minesh0111 commented:
@ishakhsuvarov, @fooman any updates on this commit ?

fooman commented:
@minesh0111 Okay I found the culprit. With your changes applied when you head to System > Index Management the Select all, etc do not work any more.

<img width="1662" alt="screen shot 2017-06-22 at 18 32 54" src="https://user-images.githubusercontent.com/455508/27420306-5f8627e8-5779-11e7-95af-5359b29bf181.png">

The problem comes from [here](https://github.com/magento/magento2/blob/develop/app/code/Magento/Indexer/view/adminhtml/layout/indexer_indexer_list_grid.xml#L20). So Magento sometimes seems to set the MassactionIdField on the parent grid block and sometimes directly on the massaction block.

One thing we could try is to change

        $massActionIdField = $this->getParentBlock()->getMassactionIdField();

to

        if ($this->getMassactionIdField()) {
            $massActionIdField = $this->getMassactionIdField();
        } else {
            $massActionIdField = $this->getParentBlock()->getMassactionIdField();
        }

and then we'll see how we go.

minesh0111 committed:
Fixed Index Management Select all

minesh0111 committed:
Fixed Index Management Select all

minesh0111 committed:
Fixed Index Management Select all

minesh0111 committed:
Fixed Index Management Select all

minesh0111 commented:
@fooman updated as suggested.

minesh0111 committed:
Travis CI Fixed

minesh0111 committed:
Travis CI Fixed

minesh0111 committed:
Travis CI Fixed

minesh0111 committed:
Travis CI Fixed

fooman commented:
Thanks @minesh0111 for getting the build green again. I'll keep you posted.

fooman commented:
Thanks @minesh0111 for persevering - your contribution has been merged.

minesh0111 commented:
Thanks @fooman, @magento-team, @convenient  and all who helped me in this contribution
